crates/dojo-core/src/auth/systems.cairo

```diff
@@ -7,100 +7,157 @@ struct Route {
-    }
-
-    fn _set_authorization_routing(ref routing: Array<Route>, world_address: starknet::ContractAddress) {
-        gas::withdraw_gas().expect('Out of gas');
-
-        if routing.is_empty() {
-            return ();
-        }
-
-        let r = routing.pop_front().unwrap();
-
-        let mut calldata = ArrayTrait::new(); 
-        serde::Serde::<Role>::serialize(@Role { id: r.role_id }, ref calldata);
-        IWorldDispatcher { contract_address: world_address }.set_entity(
-            dojo_core::string::ShortStringTrait::new('Role'), QueryTrait::new_from_id(r.target_id), 0_u8, calldata.span());
-
-        let mut calldata = ArrayTrait::new(); 
-        serde::Serde::<Status>::serialize(@Status { is_authorized: bool::True(()) }, ref calldata);
-        IWorldDispatcher { contract_address: world_address }.set_entity(
-            dojo_core::string::ShortStringTrait::new('Status'), (r.role_id, r.resource_id).into(), 0_u8, calldata.span());
+    fn execute(route: Route) {
+        // Set scoped role
+        commands::set_entity(
+            (route.target_id, route.resource_id).into(), (AuthRole { id: route.role_id })
+        );
 
-        _set_authorization_routing(ref routing, world_address);
+        // Set status
+        commands::set_entity(
+            (route.role_id, route.resource_id).into(),
+            (AuthStatus { is_authorized: bool::True(()) })
+        );
     }
 }
 
 #[system]
-mod Authorize {
+mod IsAccountAdmin {
     use traits::Into;
+    use starknet::get_tx_info;
+    use box::BoxTrait;
     use dojo_core::integer::u250;
-    use dojo_core::auth::components::Status;
-    use dojo_core::auth::components::Role;
+    use dojo_core::auth::components::AuthStatus;
+    use dojo_core::auth::components::AuthRole;
+
+    fn execute() -> bool {
+        // Get calling account contract address
+        let caller = get_tx_info().unbox().account_contract_address; // tx origin
+        let role = commands::<AuthRole>::entity(caller.into());
+        // Authorize if role is Admin
+        role.id.into() == 'Admin'
+    }
+}
 
-    fn execute(caller_id: u250, resource_id: u250) {
-        let role = commands::<Role>::entity(caller_id.into());
-        let authorization_status = commands::<Status>::entity(
-            (role.id, resource_id).into()
+#[system]
+mod IsAuthorized {
+    use traits::Into;
+    use dojo_core::integer::u250;
+    use dojo_core::auth::components::AuthStatus;
+    use dojo_core::auth::components::AuthRole;
+
+    fn execute(target_id: u250, resource_id: u250) -> bool {
+        // Get World level role
+        let maybe_role = commands::<AuthRole>::try_entity(target_id.into());
+        let role = match maybe_role {
+            Option::Some(role) => role.id.into(),
+            Option::None(()) => 0,
+        };
+
+        // Get component-scoped role
+        let maybe_scoped_role = commands::<AuthRole>::try_entity((target_id, resource_id).into());
```

Comment:
Can we order this execution logic to minimize reads? For example, I think scoped role might be the most common to begin with? Perhaps we can check and assert that first, then fall back to admin role if it isn't authorized with the scoped role?

Reply:
You're right, there's a lot of reads like this. Let me update it that